interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java

```diff
@@ -0,0 +1,231 @@
+            long arrivalTime = System.nanoTime() / 1000000;
+            Message m = new Message(arrivalTime, request, readableRequestBytes);
+            clientQueue.add(m);
+            incrementClientQSize(readableRequestBytes);
+          } else {
+            // This is arbitrary right now. Maybe sleep for some fraction of the latency?
+            Thread.sleep(1);
+          }
+        } catch (IOException e) {
+          e.printStackTrace();
+        } catch (InterruptedException e) {
+          e.printStackTrace();
+        }
+      }
+    }
+  }
+
+  private class ServerReader implements Runnable {
+
+    DataInputStream serverIn;
+
+    public ServerReader(DataInputStream serverIn) {
+      this.serverIn = serverIn;
+    }
+
+    @Override
+    public void run() {
+      while (!shutDown) {
+        try {
+          int nextChunk = Math.min(chunkSize, qLength - serverQSize);
+          if (nextChunk != 0) {
+            byte[] response = new byte[nextChunk];
+            int readableResponseBytes = serverIn.read(response);
+            // long arrivalTime = System.currentTimeMillis();
+            long arrivalTime = System.nanoTime() / 1000000;
+            Message m = new Message(arrivalTime, response, readableResponseBytes);
+            serverQueue.add(m);
+            incrementServerQSize(readableResponseBytes);
+          }
+          else {
+            Thread.sleep(1);
+          }
+
+        } catch (IOException e) {
+          e.printStackTrace();
+        } catch (InterruptedException e) {
+          e.printStackTrace();
+        }
+      }
+    }
+  }
+
+  private class ClientWriter implements Runnable {
+
+    DataOutputStream clientOut;
+
+    public ClientWriter(DataOutputStream clientOut) {
+      this.clientOut = clientOut;
+    }
+
+    @Override
+    public void run() {
+      while (!shutDown) {
+        try {
+          Message next = clientQueue.take();
+          long howLong = (next.arrivalTime + latency) - (System.nanoTime() / 1000000);
+          while (howLong > 0) {
+            Thread.sleep(howLong);
+            howLong = (next.arrivalTime + latency) - (System.nanoTime() / 1000000);
+          }
+          clientOut.write(next.message, 0, next.messageLength);
+          incrementClientQSize(-next.messageLength);
+        } catch (IOException e) {
+          e.printStackTrace();
+        } catch (InterruptedException e) {
+          e.printStackTrace();
+        }
+      }
+    }
+  }
+
+  private class ServerWriter implements Runnable {
+
+    DataOutputStream serverOut;
+
+    public ServerWriter(DataOutputStream serverOut) {
+      this.serverOut = serverOut;
+    }
+
+    @Override
+    public void run() {
+      while (!shutDown) {
+        try {
+          Message next = serverQueue.take();
+          long howLong = (next.arrivalTime + latency) - (System.nanoTime() / 1000000);
+          while (howLong > 0) {
+            Thread.sleep(howLong);
```

Comment:
You actually have to do the sleeping on the reader, not the writer. Otherwise you re-produce the issues with infinitely buffering. By avoiding reading on the client, you cause push-back on the sender very similar to if there were lost packets due to going too fast.